gadget/update_test.go

```diff
@@ -642,151 +657,1482 @@ func updateDataSet(c *C) (oldData gadget.GadgetData, newData gadget.GadgetData,
+}
+
+func (u *updateTestSuite) TestUpdateApplyUC20KernelAssetsWithInitialMapAllVolumesUpdatedFullLogic(c *C) {
+	u.restoreVolumeStructureToLocationMap()
+
+	oldKernelDir := c.MkDir()
+	newKernelDir := c.MkDir()
+
+	kernelYaml := []byte(`assets:
+  ref:
+    update: true
+    content:
+      - kernel-content
+      - kernel-content2`)
+	makeSizedFile(c, filepath.Join(newKernelDir, "meta/kernel.yaml"), 0, kernelYaml)
+	makeSizedFile(c, filepath.Join(oldKernelDir, "meta/kernel.yaml"), 0, kernelYaml)
+
+	makeSizedFile(c, filepath.Join(newKernelDir, "kernel-content"), quantity.SizeMiB, nil)
+	makeSizedFile(c, filepath.Join(newKernelDir, "kernel-content2"), quantity.SizeMiB, nil)
+	makeSizedFile(c, filepath.Join(oldKernelDir, "kernel-content"), quantity.SizeMiB, nil)
+	makeSizedFile(c, filepath.Join(oldKernelDir, "kernel-content2"), quantity.SizeMiB, nil)
+
+	oldData := gadget.GadgetData{
+		Info: &gadget.Info{
+			Volumes: map[string]*gadget.Volume{},
+		},
+		RootDir:       c.MkDir(),
+		KernelRootDir: oldKernelDir,
+	}
+
+	newData := gadget.GadgetData{
+		Info: &gadget.Info{
+			Volumes: map[string]*gadget.Volume{},
 		},
+		RootDir:       c.MkDir(),
+		KernelRootDir: newKernelDir,
 	}
-	newInfo := &gadget.Info{
-		Volumes: map[string]*gadget.Volume{
-			"foo": {
-				Bootloader: "grub",
-				Schema:     "gpt",
-				Structure:  []gadget.VolumeStructure{bareStruct, fsStruct, lastStruct},
-			},
-		},
+
+	rollbackDir := c.MkDir()
+
+	const multiVolWithKernel = `
+volumes:
+  pc:
+    schema: gpt
+    bootloader: grub
+    structure:
+      - name: mbr
+        type: mbr
+        size: 440
+      - name: BIOS Boot
+        type: DA,21686148-6449-6E6F-744E-656564454649
+        size: 1M
+        offset: 1M
+        offset-write: mbr+92
+      - name: ubuntu-seed
+        role: system-seed
+        filesystem: vfat
+        # UEFI will boot the ESP partition by default first
+        type: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+        size: 1200M
+        content:
+          - source: $kernel:ref/kernel-content
+            target: /
+      - name: ubuntu-boot
+        role: system-boot
+        filesystem: ext4
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        # whats the appropriate size?
+        size: 750M
+      - name: ubuntu-save
+        role: system-save
+        filesystem: ext4
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        size: 16M
+      - name: ubuntu-data
+        role: system-data
+        filesystem: ext4
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        size: 1G
+  foo:
+    schema: gpt
+    structure:
+      - name: barething
+        type: bare
+        size: 4096
+      - name: nofspart
+        type: A11D2A7C-D82A-4C2F-8A01-1805240E6626
+        size: 4096
+      - name: some-filesystem
+        filesystem: ext4
+        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
+        size: 1G
+        content:
+          - source: $kernel:ref/kernel-content2
```

Comment:
so the interesting case is actually when this doesn't happen

Reply:
Added a test case like this now